src/controller_examples/rabbitmq_controller/proof/liveness/helper_invariants/invariants.rs

```diff
@@ -0,0 +1,863 @@
+            && s.resource_obj_of(make_stateful_set_key(rabbitmq.object_ref())).metadata.finalizers.is_None()
+            && exists |uid: nat| #![auto]
+            s.resource_obj_of(make_stateful_set_key(rabbitmq.object_ref())).metadata.owner_references == Some(seq![OwnerReferenceView {
+                block_owner_deletion: None,
+                controller: Some(true),
+                kind: RabbitmqClusterView::kind(),
+                name: rabbitmq.metadata.name.get_Some_0(),
+                uid: uid,
+            }])
+    }
+}
+
+#[verifier(external_body)]
+pub proof fn lemma_always_stateful_set_has_no_finalizers_or_timestamp_and_only_has_controller_owner_ref(spec: TempPred<RMQCluster>, rabbitmq: RabbitmqClusterView)
+    requires
+        spec.entails(lift_state(RMQCluster::init())),
+        spec.entails(always(lift_action(RMQCluster::next()))),
+    ensures
+        spec.entails(always(lift_state(stateful_set_has_no_finalizers_or_timestamp_and_only_has_controller_owner_ref(rabbitmq)))),
+{
+    let invariant = stateful_set_has_no_finalizers_or_timestamp_and_only_has_controller_owner_ref(rabbitmq);
+    assert forall |s, s_prime: RMQCluster| invariant(s) && #[trigger] RMQCluster::next()(s, s_prime) implies invariant(s_prime) by {
+
+    }
+    init_invariant(spec, RMQCluster::init(), RMQCluster::next(), invariant);
+}
+
+pub proof fn lemma_true_leads_to_always_every_update_cm_req_does_the_same(spec: TempPred<RMQCluster>, rabbitmq: RabbitmqClusterView)
+    requires
+        valid(stable(spec)),
+        spec.entails(always(lift_state(RMQCluster::every_in_flight_msg_has_lower_id_than_allocator()))),
+        spec.entails(always(lift_action(RMQCluster::next()))),
+        spec.entails(always(lift_state(RMQCluster::busy_disabled()))),
+        spec.entails(tla_forall(|i| RMQCluster::kubernetes_api_next().weak_fairness(i))),
+        spec.entails(always(lift_state(RMQCluster::each_key_in_reconcile_is_consistent_with_its_object()))),
+        spec.entails(always(lift_state(RMQCluster::the_object_in_reconcile_has_spec_and_uid_as(rabbitmq)))),
+    ensures
+        spec.entails(
+            true_pred().leads_to(always(lift_state(every_update_cm_req_does_the_same(rabbitmq))))
+        ),
+{
+    let requirements = |msg: Message| {
+        cm_update_request_msg(rabbitmq.object_ref())(msg)
+        ==> msg.content.get_update_request().obj.spec == ConfigMapView::marshal_spec((make_server_config_map(rabbitmq).data, ()))
+    };
+    eliminate_always(spec, lift_action(RMQCluster::next()));
+    eliminate_always(spec, lift_state(RMQCluster::each_key_in_reconcile_is_consistent_with_its_object()));
+    eliminate_always(spec, lift_state(RMQCluster::the_object_in_reconcile_has_spec_and_uid_as(rabbitmq)));
+    assert forall |ex| #[trigger] spec.satisfied_by(ex) implies lift_action(RMQCluster::every_new_in_flight_req_msg_is_expected(requirements)).satisfied_by(ex) by {
```

Comment:
I feel like for this type of reasoning, we can have a rule similar to `init_invariant`, like:
```
pub proof fn init_invariant<T>(spec: TempPred<T>, init: StatePred<T>, next: ActionPred<T>, inv: ActionPred<T>)
    requires
        forall |s, s_prime: T| #[trigger] next(s, s_prime) ==> inv(s, s_prime),
        spec.entails(always(lift_action(next))),
    ensures
        spec.entails(always(lift_action(inv))),
```

Reply:
Oh I feel like this doesn't sound better. Because will this need strengthen next before calling this lemma? The stronger next thing is a nightmare to me (stronger_next => entails_and ==> always_and ==> temp_equality).

Reply:
That's fine. This is more like a personal taste/style on the proof. Strengthening next is indeed a super tedious process (unfortunately the macro is not strong enough). I am ok with that but I personally don't like exposing `ex` in the proof.

If you really prefer the current approach then you can keep it.